### PR TITLE
feat(actions): added the ability to configure actions in elgg-plugin.php

### PR DESCRIPTION
**Pull request by jdalsem**
Just want the approval to go forth with this before adding this to all the plugins

### CODE REVIEW

**Comment by jdalsem** [comment]
If this is the approach we like, i would like to do this also for widgets #10348 and plugin settings #10349

**Comment by jdalsem** [comment]
i think this will not break compatibility.... care to disagree?

**Comment by hypeJunction** [comment]
Seems reasonable. No objections from me.

**Comment by jdalsem** [comment]
updated for all plugins

**Comment by mrclay** [comment]
This shouldn't read elgg-plugin.php every request.

**Comment by mrclay** [comment]
My thinking is that elgg-plugin shouldn't just move the use of the imperative APIs, it should ideally result in caching data structures so tons of imperative function calls don't have to be made at all. Still this isn't a deal-braker for this PR as is. We can always go back and optimize how elgg-plugin gets handled.

**Comment by mrclay** [comment]
Give me a day or two with this.

File: docs/guides/actions.rst [hunk 1]
```diff
@@ -28,6 +28,26 @@ The ``mod/example/actions/example.php`` script will now be run whenever a form i
...
+You can also register actions via the :doc:`elgg-plugin</guides/plugins>` config file. 
+To do this you need to provide an action section in the config file. 
+The location of the action files are assumed to be in the plugin folder  ``/actions``.
+
+.. code:: php
+
+	<?php
+
+	return [
+		'logged_in' => [
```
Comment (mrclay) [inline docs/guides/actions.rst]: Include the actions key (make this a valid file).

Comment (jdalsem) [inline docs/guides/actions.rst]: whoops... good catch... was copying from views docs :)

**Comment by jdalsem** [comment]
> We can always go back and optimize how elgg-plugin gets handled.

yep... but with opcache this is not such a big thing at the moment or do you think it is?

**Comment by mrclay** [comment]
> with opcache this is not such a big thing at the moment

Yeah. For now, probably not worth worrying about.

**Comment by jdalsem** [comment]
addressed @mrclay's comment.. ready to go it

File: mod/bookmarks/elgg-plugin.php [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+return [
+	'actions' => [
+		'logged_in' => [
+			'bookmarks/save' => 'bookmarks/save.php',
+			'bookmarks/delete' => 'bookmarks/delete.php',
+		],
+	],
+	'views' => [
+		'default' => [
+			'bookmarks/' => __DIR__ . '/graphics',
```
Comment (mrclay) [inline mod/bookmarks/elgg-plugin.php]: @hypeJunction Editing these files with a CLI tool might be tricky due to constants like this. Just thinking ahead. Fine for now.

**Review by mrclay — APPROVED** [review]

File: mod/bookmarks/elgg-plugin.php (hunk 2 above)
Comment (hypeJunction) [inline mod/bookmarks/elgg-plugin.php]: We could define them as variables before requiring the file. Might nake sense to inject install path, root path, plgins path and root.

Comment (mrclay) [inline mod/bookmarks/elgg-plugin.php]: Unless we plan to fully parse into a tree, a string format like `"{{DIR}}/graphics"` would be easiest to work with include/var_export.

**Comment by jdalsem** [comment]
Add a new ticket for it as it is not specific for this issue

Op 8 okt. 2016 7:41 p.m. schreef "Steve Clay" <notifications@github.com>:

File: mod/bookmarks/elgg-plugin.php (hunk 2 above)
Comment (mrclay) [inline mod/bookmarks/elgg-plugin.php]: A discussion for another ticket. @jdalsem this is fine for now.

**Review by hypeJunction — COMMENTED** [review]
If we are not using an absolute path to file is there any reason to keep the filename? In 99% of cases they match the action name. I would say srop them and require that name matches filename. For core actions that don't match just do a BC include from a new location.
Also I am not sure about the benefit - it makes it harder to read the init function as you have no overview of the registered actions. It creates a precedent for moving all hook/event registrations to this format (is that something we want?). Additionally, how much longer are we going to keep action files around? We need to make a switch to using recylclable action classes eventually .

**Comment by mrclay** [comment]
I'd prefer holding this until 3.0, and figuring out how to make it simpler. As is it just allows devs to break back compatibility in their plugins. The advantages seem minimal.

**Comment by jdalsem** [comment]
> If we are not using an absolute path to file is there any reason to keep the filename? In 99% of cases they match the action name. I would say srop them and require that name matches filename. For core actions that don't match just do a BC include from a new location.

good point... can do even file renames as part of this PR if needed (if we postpone this to 3.0)



> Also I am not sure about the benefit - it makes it harder to read the init function as you have no overview of the registered actions.

they get messy over time anyway...



> It creates a precedent for moving all hook/event registrations to this format (is that something we want?).

this indeed crossed my mind... i think i want it... i even think we can drop the init events in many cases this way... and we can even drop the need to have a start.php file in your plugin



> Additionally, how much longer are we going to keep action files around? We need to make a switch to using recylclable action classes eventually .

don't think that plan is going to happen soon... probably Elgg 4? deprecating during Elgg 3



Will rebase this to 3.x and try to make registration even simpler

**Comment by jdalsem** [comment]
- rebased to master...
- made registration simpler by assuming path
- moved plugin action files to assumed location

**Review by mrclay — APPROVED** [review]

File: docs/guides/actions.rst [hunk 3]
```diff
@@ -28,6 +28,28 @@ The ``mod/example/actions/example.php`` script will now be run whenever a form i
...
+The location of the action files are assumed to be in the plugin folder  ``/actions``.
+
+.. code:: php
+
+	<?php
+
+	return [
+		'actions' => [
+			'logged_in' => [
+				'discussion/save',
```
Comment (hypeJunction) [inline docs/guides/actions.rst]: Wouldn't it be easier to read `'discussion/save' => 'logged_in'`?

Comment (mrclay) [inline docs/guides/actions.rst]: Seem equally readable to me. I like PR as we can easily move from numeric keys to `name => classname`.

File: mod/groups/elgg-plugin.php [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+			'groups/remove',
+			'groups/killrequest',
+			'groups/killinvitation',
+			'groups/addtogroup',
+		],
+		'admin' => [
+			'groups/featured',
+		],
+	],
+	'views' => [
```
Comment (hypeJunction) [inline mod/groups/elgg-plugin.php]: These are not needed anymore I think

**Review by hypeJunction — APPROVED** [review]

**Comment by jdalsem** [comment]
did a bit more thinking an was wondering if this is better:

```php
return [
    'actions' => [
        'blog/save' => [], // all defaults
        'blog/delete' => [
              'access' => 'admin',
              'priority' => 400,
              'filename' => __DIR__ . 'actions/blog/remove.php',
        ],
    ],
];
```

**Comment by hypeJunction** [comment]
That's why I think this should wait until we have a better idea of where the actions are headed to avoid API churn.

**Comment by jdalsem** [comment]
well my proposed way allows you to go everywhere after this... other approaches lock you in...

**Pull request merged**